stages/org.osbuild.coreos.live-artifacts.mono

```diff
@@ -0,0 +1,937 @@
+        shutil.copyfile(src, dst)
+        # initramfs isn't world readable by default so let's open up perms
+        os.chmod(dst, 0o644)
+
+    # Generate initramfs stamp file indicating that this is a live
+    # initramfs.  Store the build ID in it.
+    stamppath = paths["initrd/etc/coreos-live-initramfs"]
+    os.makedirs(os.path.dirname(stamppath), exist_ok=True)
+    with open(stamppath, 'w', encoding='utf8') as fh:
+        fh.write(version + '\n')
+
+    # Generate rootfs stamp file with the build ID, indicating that the
+    # rootfs has been appended and confirming that initramfs and rootfs are
+    # from the same build.
+    stamppath = os.path.join(paths["initrd-rootfs"], 'etc/coreos-live-rootfs')
+    os.makedirs(os.path.dirname(stamppath), exist_ok=True)
+    with open(stamppath, 'w', encoding='utf8') as fh:
+        fh.write(version + '\n')
+
+    # Generate JSON file that lists OS features available to
+    # coreos-installer {iso|pxe} customize.  Put it in the initramfs for
+    # pxe customize and the ISO for iso customize.
+    features = json.dumps(get_os_features(tree=deployed_tree), indent=2, sort_keys=True) + '\n'
+    featurespath = paths["initrd/etc/coreos/features.json"]
+    os.makedirs(os.path.dirname(featurespath), exist_ok=True)
+    with open(featurespath, 'w', encoding='utf8') as fh:
+        fh.write(features)
+    with open(paths["iso/coreos/features.json"], 'w', encoding='utf8') as fh:
+        fh.write(features)
+
+    mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, paths, os_release)
+
+    blsentry_kargs = mksquashfs_metal(paths, workdir, img_metal, loop_client)
+
+    # Generate rootfs image
+    # The rootfs must be uncompressed because the ISO mounts root.squashfs
+    # directly from the middle of the file
+    extend_initramfs(initramfs=paths["iso/images/pxeboot/rootfs.img"],
+                     tree=paths["initrd-rootfs"], compress=False)
+    # Check that the root.squashfs magic number is in the offset hardcoded
+    # in sysroot.mount in 20live/live-generator
+    with open(paths["iso/images/pxeboot/rootfs.img"], 'rb') as fh:
+        fh.seek(124)
+        if fh.read(4) != b'hsqs':
+            raise ValueError("root.squashfs not at expected offset in rootfs image")
+    # Save stream hash of rootfs for verifying out-of-band fetches
+    os.makedirs(os.path.dirname(paths["initrd/etc/coreos-live-want-rootfs"]), exist_ok=True)
+    make_stream_hash(paths["iso/images/pxeboot/rootfs.img"],
+                     paths["initrd/etc/coreos-live-want-rootfs"])
+    # Add common content
+    extend_initramfs(initramfs=paths["iso/images/pxeboot/initrd.img"], tree=paths["initrd"])
+
+    filenames = options['filenames']
+    gen_live_artifacts(paths, tree, filenames, deployed_tree, loop_client, version, blsentry_kargs)
+
+
+if __name__ == "__main__":
+    args = osbuild.api.arguments()
+    with tempfile.TemporaryDirectory(dir=args["tree"]) as tmdir:
+        r = main(workdir=tmdir,
```

Comment:
AFAICT we never return anything from main() so this r and the subsequent `sys.exit(r)` seems unnecessary (i.e. main() either works and just reaches the end of the function which means an implicit "return None" or it errors with an exception

Reply:
I think I was just following what is done in org.osbuild.bootiso.mono

Reply:
maybe let's fix this in a followup that changes both stages to something more correct.